storage: add Allocator.TransferLease{Source,Target}

petermattis opened the pull request:
Before (`allocsim -n 4 -w 4`):

```
_elapsed__ops/sec___errors_replicas_________1_________2_________3_________4
    5m0s   1448.8        0     1686   559/554     383/0     331/1     413/2
```

After:

```
_elapsed__ops/sec___errors_replicas_________1_________2_________3_________4
    5m0s   1413.7        0     1658   406/132   422/141   410/136   420/138
```

I do see a little bit of thrashiness with the lease-transfer heuristics,
though it settles down relatively quickly so I'm not sure if it is worth
addressing yet.

Depends on #10420

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10464)
<!-- Reviewable:end -->

petermattis committed:
storage: add Allocator.TransferLease{Source,Target}

Before (`allocsim -n 4 -w 4`):

_elapsed__ops/sec___errors_replicas_________1_________2_________3_________4
    5m0s   1472.5        0     2124   704/687     492/0     522/0     406/0

After:

_elapsed__ops/sec___errors_replicas_________1_________2_________3_________4
    5m0s   1506.7        0     2157   520/183   548/180   547/166   542/180

I do see a little bit of thrashiness with the lease-transfer heuristics,
though it settles down relatively quickly so I'm not sure if it is worth
addressing yet.

Depends on #10420

tamird commented:
Could you look at https://github.com/cockroachdb/cockroach/issues/10193 and https://github.com/cockroachdb/cockroach/issues/10497 before moving ahead with this? There's something funky going on with lease transfers in the latter (and the former has a similar issue).

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 379 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVydVil_Or2w9RqZD30:-KVydVil_Or2w9RqZD31:b42j5cz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/allocator.go#L379)):*
> ```Go
> }
> 
> // TransferLeaseSource returns true if the specified store is overfull with
> ```

can you clarify that overfull refers to leases here? also this is a strange name for a method that returns a bool

---

*[pkg/storage/allocator_test.go, line 703 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVygvhMMUExBJEoOxjD:-KVygvhMMUExBJEoOxjE:b-dbf1o7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/allocator_test.go#L703)):*
> ```Go
> 	sg.GossipStores(stores, t)
> 
> 	makeReplicaDescs := func(storeIDs ...roachpb.StoreID) []roachpb.ReplicaDescriptor {
> ```

how come each test case needs a new slice?

---

*[pkg/storage/allocator_test.go, line 734 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyhB-3VqF7MxzL7TdG:-KVyhB-3VqF7MxzL7TdH:bdkkmt0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/allocator_test.go#L734)):*
> ```Go
> 		target := a.TransferLeaseTarget(config.Constraints{}, c.existing, c.leaseholder, 0, c.check)
> 		if c.expected != target.StoreID {
> 			t.Fatalf("%d: expected %d, but found %d", i, c.expected, target.StoreID)
> ```

optional: could be nice to introduce sub-tests here, which would give each test a name

---

*[pkg/storage/allocator_test.go, line 768 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyhEbQYqrFz3su8e7F:-KVyhEbQYqrFz3su8e7G:bdkkmt0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/allocator_test.go#L768)):*
> ```Go
> 		result := a.TransferLeaseSource(config.Constraints{}, c.leaseholder, 0)
> 		if c.expected != result {
> 			t.Fatalf("%d: expected %v, but found %v", i, c.expected, result)
> ```

optional: could be nice to introduce sub-tests here, which would give each test a name

---

*[pkg/storage/replicate_queue.go, line 188 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyeCfhi0cJ-FQDTJVq:-KVyeCfinnMvrnHFBttZ:b-x6ndlc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/replicate_queue.go#L188)):*
> ```Go
> 	case AllocatorRemove:
> 		log.Event(ctx, "removing a replica")
> 		removeReplica, err := rq.allocator.RemoveTarget(desc.Replicas, 0)
> ```

what does zero mean here? this is the only "real" caller to `RemoveTarget`, so this change in semantics should probably be achieved by removing the second argument completely.

---

*[pkg/storage/replicate_queue.go, line 192 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyenu91ENNy_Lhd-f4:-KVyenu91ENNy_Lhd-f5:b-kc0a3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/replicate_queue.go#L192)):*
> ```Go
> 			return err
> 		}
> 		// We require the lease in order to process replicas, so
> ```

what is this comment now explaining?

---

*[pkg/storage/replicate_queue.go, line 207 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyf3Kbxr5dD_su2LFN:-KVyf3Kbxr5dD_su2LFO:b-btgvu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/replicate_queue.go#L207)):*
> ```Go
> 				zone.Constraints, desc.Replicas, repl.store.StoreID(), desc.RangeID,
> 				false /* check-transfer-lease-source */)
> 			if target.StoreID != 0 {
> ```

nit: `if (target != roachpb.ReplicaDescriptor{}) {`

---

*[pkg/storage/replicate_queue.go, line 210 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyfW1PO11xkQSLACUp:-KVyfW1PO11xkQSLACUq:bwofqpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/replicate_queue.go#L210)):*
> ```Go
> 				log.VEventf(ctx, 1, "transferring lease to s%d", target.StoreID)
> 				if err := repl.AdminTransferLease(target.StoreID); err != nil {
> 					return errors.Wrapf(err, "%s: unable to transfer lease", repl)
> ```

should this include the store to which the lease transfer was sent?

---

*[pkg/storage/replicate_queue.go, line 247 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyfjZ0h0i2W89AeEfx:-KVyfjZ1PXCcQ6xNqFuk:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/replicate_queue.go#L247)):*
> ```Go
> 			log.VEventf(ctx, 1, "transferring lease to s%d", target.StoreID)
> 			if err := repl.AdminTransferLease(target.StoreID); err != nil {
> 				return errors.Wrapf(err, "%s: unable to transfer lease", repl)
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-:-KVyfrdx-s9uwxcNCqGM:bk6i37b)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I can't reproduce the failures in #10193 and #10497 due to another error occurring first. See #10501. 

---

Review status: 0 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/storage/allocator.go, line 379 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVydVil_Or2w9RqZD30:-KVzP7Ix4o8grSOi8CH0:b-dl5doo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/allocator.go#L379)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you clarify that overfull refers to leases here? also this is a strange name for a method that returns a bool
</details>
Clarified comment. Renamed to `ShouldTransferLease`.

---

*[pkg/storage/allocator_test.go, line 703 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVygvhMMUExBJEoOxjD:-KVzPKs3DP3BZdQ89-Fr:bbh9ve1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/allocator_test.go#L703)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come each test case needs a new slice?
</details>
They don't. Changed.

---

*[pkg/storage/allocator_test.go, line 734 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyhB-3VqF7MxzL7TdG:-KVzPrruOYFb2YQ5f-R8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/allocator_test.go#L734)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> optional: could be nice to introduce sub-tests here, which would give each test a name
</details>
Done.

---

*[pkg/storage/allocator_test.go, line 768 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyhEbQYqrFz3su8e7F:-KVzPsZohtAvv_6shTWF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/allocator_test.go#L768)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> optional: could be nice to introduce sub-tests here, which would give each test a name
</details>
Done.

---

*[pkg/storage/replicate_queue.go, line 188 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyeCfhi0cJ-FQDTJVq:-KVzPuWV7Jxpa3jELczn:b-v05t3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/replicate_queue.go#L188)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what does zero mean here? this is the only "real" caller to `RemoveTarget`, so this change in semantics should probably be achieved by removing the second argument completely.
</details>
Removed the second argument.

---

*[pkg/storage/replicate_queue.go, line 192 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyenu91ENNy_Lhd-f4:-KVzQEBEj73ftwZ0T8WJ:be1ssf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/replicate_queue.go#L192)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what is this comment now explaining?
</details>
Added a local which makes it more obvious that `repl.store.StoreID()` is the `leaseholderStoreID`.

---

*[pkg/storage/replicate_queue.go, line 207 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyf3Kbxr5dD_su2LFN:-KVzQSierYOTgZUfLA4a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/replicate_queue.go#L207)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: `if (target != roachpb.ReplicaDescriptor{}) {`
</details>
Done.

---

*[pkg/storage/replicate_queue.go, line 210 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyfW1PO11xkQSLACUp:-KVzQWtbIoyW_7OYHRZA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/replicate_queue.go#L210)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> should this include the store to which the lease transfer was sent?
</details>
Done.

---

*[pkg/storage/replicate_queue.go, line 247 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVyfjZ0h0i2W89AeEfx:-KVzQZvcZthc__j0l_xu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/replicate_queue.go#L247)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-:-KVzQaAzMLg0pqW09uk6:bu5m0s8)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 239 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVzTWg5ZpvvohCwJM_y:-KVzTWg5ZpvvohCwJM_z:b3vz63m) ([raw file](https://github.com/cockroachdb/cockroach/blob/316ed6559b3c406bf2d313335af48e74a9bf6a6b/pkg/storage/allocator.go#L239)):*
> ```Go
> 
> // RemoveTarget returns a suitable replica to remove from the provided replica
> // set. It first attempts to randomly select a target from the set nodes that
> ```

set _of_ nodes?

---

*[pkg/storage/allocator_test.go, line 703 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVygvhMMUExBJEoOxjD:-KVzTiRHKe_7Lk1y27f1:b-z0oruf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/allocator_test.go#L703)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> They don't. Changed.
</details>
so why keep this function?

---

*[pkg/storage/allocator_test.go, line 737 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVzTxEbDRxbNddIMXQC:-KVzTxEbDRxbNddIMXQD:b3asm5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/316ed6559b3c406bf2d313335af48e74a9bf6a6b/pkg/storage/allocator_test.go#L737)):*
> ```Go
> 				c.existing, c.leaseholder, 0, c.check)
> 			if c.expected != target.StoreID {
> 				t.Fatalf("%d: expected %d, but found %d", i, c.expected, target.StoreID)
> ```

the integer here is now duplicating the test name. also, i was imagining that the names would take the place of the comments you have on each test case rather than be integers.

---

*[pkg/storage/allocator_test.go, line 773 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVzTotPjdsJk0KhcGLa:-KVzTotPjdsJk0KhcGLb:b3asm5g) ([raw file](https://github.com/cockroachdb/cockroach/blob/316ed6559b3c406bf2d313335af48e74a9bf6a6b/pkg/storage/allocator_test.go#L773)):*
> ```Go
> 			result := a.ShouldTransferLease(config.Constraints{}, c.leaseholder, 0)
> 			if c.expected != result {
> 				t.Fatalf("%d: expected %v, but found %v", i, c.expected, result)
> ```

the integer here is now duplicating the test name. also, i was imagining that the names would take the place of the comments you have on each test case rather than be integers.

---

*[pkg/storage/allocator_test.go, line 840 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVzU-EdGHI6IJTeqVZp:-KVzU-Ees5dqag2TM260:b-t3w6id) ([raw file](https://github.com/cockroachdb/cockroach/blob/316ed6559b3c406bf2d313335af48e74a9bf6a6b/pkg/storage/allocator_test.go#L840)):*
> ```Go
> 	for i := 0; i < 100; i++ {
> 		// Stores 2 and 3 are overfull, so we should see them randomly selected as
> 		// the removal target..
> ```

double period

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-:-KVzUGJf7wFdo-Q_QBIc:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 239 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVzTWg5ZpvvohCwJM_y:-KVzVOXuO_UcKLZ0zOB6:bjc1xpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/316ed6559b3c406bf2d313335af48e74a9bf6a6b/pkg/storage/allocator.go#L239)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> set _of_ nodes?
</details>
`set of stores`

---

*[pkg/storage/allocator_test.go, line 703 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVygvhMMUExBJEoOxjD:-KVzVb3NxcRsylp9CrE8:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9650a585b789789426dadda3e9b2ed1c639f136b/pkg/storage/allocator_test.go#L703)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> so why keep this function?
</details>
Removed.

---

*[pkg/storage/allocator_test.go, line 737 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVzTxEbDRxbNddIMXQC:-KVzVxdeycSicT1YJOzZ:b374i78) ([raw file](https://github.com/cockroachdb/cockroach/blob/316ed6559b3c406bf2d313335af48e74a9bf6a6b/pkg/storage/allocator_test.go#L737)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> the integer here is now duplicating the test name. also, i was imagining that the names would take the place of the comments you have on each test case rather than be integers.
</details>
I don't have names for all of the test cases. Feels burdensome to require that. The integer at least allows the running of a particular test case easily.

---

*[pkg/storage/allocator_test.go, line 773 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVzTotPjdsJk0KhcGLa:-KVzW5MRoeTsF3P6kPbm:b-by9q0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/316ed6559b3c406bf2d313335af48e74a9bf6a6b/pkg/storage/allocator_test.go#L773)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> the integer here is now duplicating the test name. also, i was imagining that the names would take the place of the comments you have on each test case rather than be integers.
</details>
See other comment.

---

*[pkg/storage/allocator_test.go, line 840 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVzU-EdGHI6IJTeqVZp:-KVzW6kPiZ_UIMyQm8jL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/316ed6559b3c406bf2d313335af48e74a9bf6a6b/pkg/storage/allocator_test.go#L840)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> double period
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/allocator_test.go, line 737 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVzTxEbDRxbNddIMXQC:-KVz_TsC6ei7KewXldoy:bmco6dz) ([raw file](https://github.com/cockroachdb/cockroach/blob/316ed6559b3c406bf2d313335af48e74a9bf6a6b/pkg/storage/allocator_test.go#L737)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I don't have names for all of the test cases. Feels burdensome to require that. The integer at least allows the running of a particular test case easily.
</details>
OK. Still duplicating it in the error message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/storage/allocator_test.go, line 737 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KVzTxEbDRxbNddIMXQC:-KVzbLaG7D1g07jaGK5w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/316ed6559b3c406bf2d313335af48e74a9bf6a6b/pkg/storage/allocator_test.go#L737)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> OK. Still duplicating it in the error message.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 4 files at r2, 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 334 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW1XXJ8W4tfneO4A3DD:-KW1XXJ8W4tfneO4A3DE:brukwdd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1cbdff312b51070a4bb63abe2537bd92f4d1050/pkg/storage/allocator.go#L334)):*
> ```Go
> // to from the provided list. It excludes the current lease holder replica.
> func (a *Allocator) TransferLeaseTarget(
> 	constraints config.Constraints,
> ```

Why does this take `constraints` but `RemoveTarget` doesn't? Both are acting on the current replicas of the range and comparing to an average, so it seems like either both should use the constraints to compute that average or neither should.

---

*[pkg/storage/allocator_test.go, line 729 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW1ZZPX14aX7SmNMRWB:-KW1ZZPX14aX7SmNMRWC:b-yh1vui) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1cbdff312b51070a4bb63abe2537bd92f4d1050/pkg/storage/allocator_test.go#L729)):*
> ```Go
> 	for i, c := range testCases {
> 		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
> 			target := a.TransferLeaseTarget(config.Constraints{},
> ```

We should have some tests with non-empty constraints too (doesn't have to be in this PR)

---

*[pkg/storage/replicate_queue.go, line 207 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW1_Uj9URHneJHHzshr:-KW1_Uj9URHneJHHzshs:b-lu4iki) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1cbdff312b51070a4bb63abe2537bd92f4d1050/pkg/storage/replicate_queue.go#L207)):*
> ```Go
> 			target := rq.allocator.TransferLeaseTarget(
> 				zone.Constraints, desc.Replicas, repl.store.StoreID(), desc.RangeID,
> 				false /* check-transfer-lease-source */)
> ```

Nit: either remove the hyphens in this comment (and the one below) or change it to the go-style variable name. 

---

*[pkg/storage/replicate_queue.go, line 242 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW1_rKrOvy2kzCA8wrX:-KW1_rKrOvy2kzCA8wrY:b-g3obka) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1cbdff312b51070a4bb63abe2537bd92f4d1050/pkg/storage/replicate_queue.go#L242)):*
> ```Go
> 		// We require the lease in order to process replicas, so
> 		// repl.store.StoreID() corresponds to the lease-holder's store ID.
> 		target := rq.allocator.TransferLeaseTarget(
> ```

There are two behavioral changes in this PR: the fact that AllocatorRemove will now transfer leases away from overfull stores, and this one to balance leases in AllocatorNoop. What is the result of allocsim with this part commented out (to see the first part in isolation)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-:-KW1a3IqhJfCUeygfgxh:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 334 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW1XXJ8W4tfneO4A3DD:-KW3R722nU1ARwzMDAk5:b-ya25cj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1cbdff312b51070a4bb63abe2537bd92f4d1050/pkg/storage/allocator.go#L334)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why does this take `constraints` but `RemoveTarget` doesn't? Both are acting on the current replicas of the range and comparing to an average, so it seems like either both should use the constraints to compute that average or neither should.
</details>
`RemoveTarget` will be taking constraints soon too. I didn't want to add that here because it is being added in Bram's open PR to re-add the rule-based allocator.

---

*[pkg/storage/allocator_test.go, line 729 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW1ZZPX14aX7SmNMRWB:-KW3RIIywz99NWNmtFlz:b-md4frx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1cbdff312b51070a4bb63abe2537bd92f4d1050/pkg/storage/allocator_test.go#L729)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> We should have some tests with non-empty constraints too (doesn't have to be in this PR)
</details>
TODO added.

---

*[pkg/storage/replicate_queue.go, line 188 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW3UZrOccnEIvk2JqWU:-KW3UZrOccnEIvk2JqWV:b-b5axox) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1cbdff312b51070a4bb63abe2537bd92f4d1050/pkg/storage/replicate_queue.go#L188)):*
> ```Go
> 	case AllocatorRemove:
> 		log.Event(ctx, "removing a replica")
> 		removeReplica, err := rq.allocator.RemoveTarget(desc.Replicas)
> ```

Btw, I removed the `leaseStoreID` parameter, but I think I should add it back and only transfer the lease on removal if the leaseHolder is "overfull". 

---

*[pkg/storage/replicate_queue.go, line 207 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW1_Uj9URHneJHHzshr:-KW3Rc6mWwtK-JBYd3UK:b-5k1t45) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1cbdff312b51070a4bb63abe2537bd92f4d1050/pkg/storage/replicate_queue.go#L207)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Nit: either remove the hyphens in this comment (and the one below) or change it to the go-style variable name. 
</details>
Camel-cased the name in the comment.

---

*[pkg/storage/replicate_queue.go, line 242 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW1_rKrOvy2kzCA8wrX:-KW3TpeI1ylXcRerMBX1:bxizvyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1cbdff312b51070a4bb63abe2537bd92f4d1050/pkg/storage/replicate_queue.go#L242)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> There are two behavioral changes in this PR: the fact that AllocatorRemove will now transfer leases away from overfull stores, and this one to balance leases in AllocatorNoop. What is the result of allocsim with this part commented out (to see the first part in isolation)?
</details>
```
_elapsed__ops/sec___errors_replicas_________1_________2_________3_________4
    5m0s   1690.3        0     2653   663/250   647/222   675/175   668/227
```

I ran this on a different machine, so ignore the total number of replicas and ops/sec numbers. Without active lease rebalancing we have nodes that are +13%, +0.5%, -20% and +3% from the mean. That isn't horrible (certainly better than what we have now), but I think the complication of active lease rebalancing is worthwhile as it keeps all nodes within 5% of the mean. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/allocator.go, line 243 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW3bh71NSZCg3hnP59o:-KW3bh71NSZCg3hnP59p:b-vft4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/af4be7d83da68beefbfeb7c20ca5bf09956032a2/pkg/storage/allocator.go#L243)):*
> ```Go
> // falls back to selecting a random target from any of the existing
> // replicas. It also will exclude any replica that belongs to the range lease
> // holder's store ID if leaseStoreID is non-zero.
> ```

nit: this function doesn't actually check if leaseStoreID is non-zero

---

*[pkg/storage/allocator.go, line 257 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW3bTJOA4CxwZz46KNe:-KW3bTJPocP_mG1Z5ctW:bna39jy) ([raw file](https://github.com/cockroachdb/cockroach/blob/af4be7d83da68beefbfeb7c20ca5bf09956032a2/pkg/storage/allocator.go#L257)):*
> ```Go
> 
> 	// Retrieve store descriptors for the provided replicas from the StorePool.
> 	var descriptors []roachpb.StoreDescriptor
> ```

NYC but could pre-allocate here

---

*[pkg/storage/allocator_test.go, line 729 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW1ZZPX14aX7SmNMRWB:-KW3bYNnDQsfWVPICZls:b-ti4np9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1cbdff312b51070a4bb63abe2537bd92f4d1050/pkg/storage/allocator_test.go#L729)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> TODO added.
</details>
Let's file an issue when this merges.

---

*[pkg/storage/replicate_queue.go, line 188 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW3c1GqjbrD1qSBgKuH:-KW3c1GqjbrD1qSBgKuI:b1i04hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/af4be7d83da68beefbfeb7c20ca5bf09956032a2/pkg/storage/replicate_queue.go#L188)):*
> ```Go
> 	case AllocatorRemove:
> 		log.Event(ctx, "removing a replica")
> 		// If the lease holder (our local store) is on an overfull store (in terms
> ```

"our local store is on an overfull store" - can you improve this phrasing?

---

*[pkg/storage/replicate_queue.go, line 192 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW3cJuktxGmtzFo-f7K:-KW3cJulKN6RDvtmMqoj:brr2it9) ([raw file](https://github.com/cockroachdb/cockroach/blob/af4be7d83da68beefbfeb7c20ca5bf09956032a2/pkg/storage/replicate_queue.go#L192)):*
> ```Go
> 		leaseHolderStoreID := repl.store.StoreID()
> 		if rq.allocator.ShouldTransferLease(zone.Constraints, leaseHolderStoreID, desc.RangeID) {
> 			leaseHolderStoreID = 0
> ```

i'm confused about this logic. as I understand it, this is doing:
```
considerLeaseHolder := ShouldTransferLease(leaseHolder)
RemoveTarget(..., considerLeaseHolder)
```
But doesn't RemoveTarget already consider all the replicas and only suggest removal of replicas which are on overfull stores? ISTM that the previous logic (in r4) was exactly equivalent, and AFAICT no tests changed to support this change except for a unit test of RemoveTarget which doesn't demonstrate a meaningful change in the allocator's behaviour.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/allocator.go, line 243 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW3bh71NSZCg3hnP59o:-KW3ddD8cKG4e9n-HVgs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af4be7d83da68beefbfeb7c20ca5bf09956032a2/pkg/storage/allocator.go#L243)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: this function doesn't actually check if leaseStoreID is non-zero
</details>
Done.

---

*[pkg/storage/allocator.go, line 257 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW3bTJOA4CxwZz46KNe:-KW3dgSw2Z6-TfVSsBAz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af4be7d83da68beefbfeb7c20ca5bf09956032a2/pkg/storage/allocator.go#L257)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> NYC but could pre-allocate here
</details>
Done.

---

*[pkg/storage/replicate_queue.go, line 188 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW3c1GqjbrD1qSBgKuH:-KW3dl7uq42xMWfJYR0B:bz7mygz) ([raw file](https://github.com/cockroachdb/cockroach/blob/af4be7d83da68beefbfeb7c20ca5bf09956032a2/pkg/storage/replicate_queue.go#L188)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "our local store is on an overfull store" - can you improve this phrasing?
</details>
`s/on an/an/g`

---

*[pkg/storage/replicate_queue.go, line 192 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW3cJuktxGmtzFo-f7K:-KW3doZFJbRoSAWHUvsC:b-y8xnol) ([raw file](https://github.com/cockroachdb/cockroach/blob/af4be7d83da68beefbfeb7c20ca5bf09956032a2/pkg/storage/replicate_queue.go#L192)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i'm confused about this logic. as I understand it, this is doing:
> ```
> considerLeaseHolder := ShouldTransferLease(leaseHolder)
> RemoveTarget(..., considerLeaseHolder)
> ```
> But doesn't RemoveTarget already consider all the replicas and only suggest removal of replicas which are on overfull stores? ISTM that the previous logic (in r4) was exactly equivalent, and AFAICT no tests changed to support this change except for a unit test of RemoveTarget which doesn't demonstrate a meaningful change in the allocator's behaviour.
</details>
`RemoveTarget` tries to remove from stores with too many replicas. This is allowing the leaseholder to only be considered a candidate for removal if it has too many leases.

We don't have any direct tests of the `replicateQueue`. After #10515 goes in, I'll add one for basic lease balancing. But even then it will be difficult to adequately exercise this logic in a unit test environment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replicate_queue.go, line 192 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW3cJuktxGmtzFo-f7K:-KW3iNlJWvn5r4SWzjpm:bhf3ygh) ([raw file](https://github.com/cockroachdb/cockroach/blob/af4be7d83da68beefbfeb7c20ca5bf09956032a2/pkg/storage/replicate_queue.go#L192)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `RemoveTarget` tries to remove from stores with too many replicas. This is allowing the leaseholder to only be considered a candidate for removal if it has too many leases.
> 
> We don't have any direct tests of the `replicateQueue`. After #10515 goes in, I'll add one for basic lease balancing. But even then it will be difficult to adequately exercise this logic in a unit test environment.
</details>
Ah, I see. Thanks.

The logic surrounding these allocator decisions feels like it's getting more complex to reason about with the addition of lease rebalancing as a result of the implementation strategy where we have a family of methods for each heuristic, and one is forced to page-in both families and their interactions in order to understand what's going on.

Perhaps this is just because I haven't been in the code (only reviews), but my intuition is that this stuff would be much easier to reason about if there was just a single family of methods that internally dealt with however many heuristics were involved (helpers that calculate scores, etc. are fine, obviously).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replicate_queue.go, line 192 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW3cJuktxGmtzFo-f7K:-KW3jfVkk5cg2-r_iWVw:b-71jmwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/af4be7d83da68beefbfeb7c20ca5bf09956032a2/pkg/storage/replicate_queue.go#L192)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Ah, I see. Thanks.
> 
> The logic surrounding these allocator decisions feels like it's getting more complex to reason about with the addition of lease rebalancing as a result of the implementation strategy where we have a family of methods for each heuristic, and one is forced to page-in both families and their interactions in order to understand what's going on.
> 
> Perhaps this is just because I haven't been in the code (only reviews), but my intuition is that this stuff would be much easier to reason about if there was just a single family of methods that internally dealt with however many heuristics were involved (helpers that calculate scores, etc. are fine, obviously).
</details>
Yeah, it's unfortunate how much logic is contained in `replicateQueue.process`. I'm reluctant to do anything in the short term with the rule-based allocator PR in flight. Let me think about this for a future PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replicate_queue_test.go, line 98 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW46ZX1Y2p9CNHrzKTd:-KW46ZX1Y2p9CNHrzKTe:bdopbhk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e109d2e3f2bea129bf81b5a91df69871269cb42e/pkg/storage/replicate_queue_test.go#L98)):*
> ```Go
> 		counts := countReplicas()
> 		for _, c := range counts {
> 			if c < minReplicas {
> ```

@tamird Notice the improvement that was possible here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replicate_queue_test.go, line 67 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW47RBKUbSxSzy7uN0w:-KW47RBKUbSxSzy7uN0x:blk1l0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e109d2e3f2bea129bf81b5a91df69871269cb42e/pkg/storage/replicate_queue_test.go#L67)):*
> ```Go
> 		for {
> 			if _, _, err := tc.SplitRange(splitKey); err != nil {
> 				if testutils.IsError(err, "split at key .* failed: conflict updating range descriptors") ||
> ```

can able be written as `IsError("range is already split at key|plit at key .* failed: conflict updating range descriptors")`. feel free to ignore

---

*[pkg/storage/replicate_queue_test.go, line 91 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW47Fj2anTGFjPdIhJd:-KW47Fj2anTGFjPdIhJe:beoa4le) ([raw file](https://github.com/cockroachdb/cockroach/blob/e109d2e3f2bea129bf81b5a91df69871269cb42e/pkg/storage/replicate_queue_test.go#L91)):*
> ```Go
> 	}
> 
> 	const numRanges = 15
> ```

isn't this susceptible to rot? It'd be nice to count them.

---

*[pkg/storage/replicate_queue_test.go, line 98 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW46ZX1Y2p9CNHrzKTd:-KW47ZHHLtwdtHv0ejur:b-wdj8z5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e109d2e3f2bea129bf81b5a91df69871269cb42e/pkg/storage/replicate_queue_test.go#L98)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> @tamird Notice the improvement that was possible here.
</details>
Nice, thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Added `COCKROACH_ENABLE_LEASE_REBALANCING` env var which defaults to true.

petermattis commented:
Review status: 4 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replicate_queue_test.go, line 67 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW47RBKUbSxSzy7uN0w:-KW4BlaRTWrbXjC21L0_:b-belbgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e109d2e3f2bea129bf81b5a91df69871269cb42e/pkg/storage/replicate_queue_test.go#L67)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can able be written as `IsError("range is already split at key|plit at key .* failed: conflict updating range descriptors")`. feel free to ignore
</details>
Yeah, but that makes the line really long.

---

*[pkg/storage/replicate_queue_test.go, line 91 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW47Fj2anTGFjPdIhJd:-KW4BnHWN7JyfH6cb8Gd:byy4ywk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e109d2e3f2bea129bf81b5a91df69871269cb42e/pkg/storage/replicate_queue_test.go#L91)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> isn't this susceptible to rot? It'd be nice to count them.
</details>
Sure, though if you change numbers above the test will start to fail.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replicate_queue_test.go, line 91 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW47Fj2anTGFjPdIhJd:-KW4DL32wgwOae-9XpRl:b-6furel) ([raw file](https://github.com/cockroachdb/cockroach/blob/e109d2e3f2bea129bf81b5a91df69871269cb42e/pkg/storage/replicate_queue_test.go#L91)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Sure, though if you change numbers above the test will start to fail.
</details>
What do you mean? What if a new system table is added? The test will start to fail then, too. Surely that'd be good to avoid?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 4 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replicate_queue_test.go, line 91 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW47Fj2anTGFjPdIhJd:-KW4OXJ5HM8F7Y_T6iyj:b-c1xzzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e109d2e3f2bea129bf81b5a91df69871269cb42e/pkg/storage/replicate_queue_test.go#L91)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> What do you mean? What if a new system table is added? The test will start to fail then, too. Surely that'd be good to avoid?
</details>
Done, though stress testing revealed this is now flaky (it was flaky with the hardcoded constants too). One of the flakes should be fixed by #10530. Another is #10501.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replicate_queue_test.go, line 91 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW47Fj2anTGFjPdIhJd:-KW4QUy_o1f2IDHXwbcS:b-7ukqxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e109d2e3f2bea129bf81b5a91df69871269cb42e/pkg/storage/replicate_queue_test.go#L91)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Done, though stress testing revealed this is now flaky (it was flaky with the hardcoded constants too). One of the flakes should be fixed by #10530. Another is #10501.
</details>
Was it found by the stress testing I added?! :heart_eyes_cat: 

---

*[pkg/storage/replicate_queue_test.go, line 54 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW4QS24tTE8JBNkkG8-:-KW4QS24tTE8JBNkkG80:b-v2areg) ([raw file](https://github.com/cockroachdb/cockroach/blob/adafe5960246cb7b9e93ad7cc05255c1c51f88d0/pkg/storage/replicate_queue_test.go#L54)):*
> ```Go
> 	//
> 	//   replicate_queue_test.go:71: /Table/53/0: lookup range unexpected error:
> 	//   replica corruption (processed=true): range does not match splits:
> ```

this looks like...the range just did not split? scary indeed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 4 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replicate_queue_test.go, line 91 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW47Fj2anTGFjPdIhJd:-KWE76mhO0NeoStjpu18:b-9ka1ba) ([raw file](https://github.com/cockroachdb/cockroach/blob/e109d2e3f2bea129bf81b5a91df69871269cb42e/pkg/storage/replicate_queue_test.go#L91)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Was it found by the stress testing I added?! :heart_eyes_cat: 
</details>
No, just good old fashion manual stress testing.

---

*[pkg/storage/replicate_queue_test.go, line 54 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW4QS24tTE8JBNkkG8-:-KWE71hy-90BLeF0sqY3:bkb8iwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/adafe5960246cb7b9e93ad7cc05255c1c51f88d0/pkg/storage/replicate_queue_test.go#L54)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this looks like...the range just did not split? scary indeed.
</details>
Yes indeed. Unfortunately I haven't seen this error again.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replicate_queue_test.go, line 54 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW4QS24tTE8JBNkkG8-:-KWE7riEtcstMGq8dL67:b-v09neg) ([raw file](https://github.com/cockroachdb/cockroach/blob/adafe5960246cb7b9e93ad7cc05255c1c51f88d0/pkg/storage/replicate_queue_test.go#L54)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yes indeed. Unfortunately I haven't seen this error again.
</details>
did you mean to set this back to 10, or why remove the comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replicate_queue_test.go, line 54 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KW4QS24tTE8JBNkkG8-:-KWE8EIaZa8xijdkmkb-:banesc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/adafe5960246cb7b9e93ad7cc05255c1c51f88d0/pkg/storage/replicate_queue_test.go#L54)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> did you mean to set this back to 10, or why remove the comment?
</details>
I ran into a variety of flakiness problems with 10 nodes. Worth tracking down, but not holding up this PR. I'll file an issue. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Tracked down the `TestReplicateQueueRebalance` flakiness which was necessitated a small tweak to `replicateQueue.shouldQueue`. PTAL.

tamird commented:
Reviewed 4 of 4 files at r11.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 556 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KWYcP9dyRtIarNu2Yyk:-KWYcP9dyRtIarNu2Yyl:b52kycu) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc2f9d160dd938d3ac8b0824abde1d0afd108d06/pkg/storage/allocator.go#L556)):*
> ```Go
> var enableLeaseRebalancing = envutil.EnvOrDefaultBool("COCKROACH_ENABLE_LEASE_REBALANCING", true)
> 
> func shouldTransferLease(sl StoreList, candidate roachpb.StoreDescriptor) bool {
> ```

s/candidate/source/g for consistency with the caller. perhaps this should be a method on the store list?

---

*[pkg/storage/replicate_queue.go, line 117 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KWYc2CMi7-k7DxnC0vP:-KWYc2CMi7-k7DxnC0vQ:ba5pahb) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc2f9d160dd938d3ac8b0824abde1d0afd108d06/pkg/storage/replicate_queue.go#L117)):*
> ```Go
> 		return true, priority
> 	}
> 	// See if there is a rebalancing opportunity present. Note that we don't
> ```

"see if there's a rebalancing opportunity present" is not a useful comment.

The rest of this comment seems pretty complicated - it talks about the negative case (no current lease) rather than the positive case which the code that immediately follows it operates on. It seems to me that if there _is_ a current lease, then the lease holder is immediately considered for lease rebalancing, and that's all that this comment really needs to say.

Also, is it possible to get here without holding the lease? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 556 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KWYcP9dyRtIarNu2Yyk:-KWYetTL6MyrGx4igiE1:b-5gecd2) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc2f9d160dd938d3ac8b0824abde1d0afd108d06/pkg/storage/allocator.go#L556)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/candidate/source/g for consistency with the caller. perhaps this should be a method on the store list?
</details>
Done.

I don't see a reason to prefer making this a method on `StoreList`. I don't see an argument against doing so either. I'm going to leave as-is for now.

---

*[pkg/storage/replicate_queue.go, line 117 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KWYc2CMi7-k7DxnC0vP:-KWYfHZsI0MV36tQQsAw:bugtc4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc2f9d160dd938d3ac8b0824abde1d0afd108d06/pkg/storage/replicate_queue.go#L117)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "see if there's a rebalancing opportunity present" is not a useful comment.
> 
> The rest of this comment seems pretty complicated - it talks about the negative case (no current lease) rather than the positive case which the code that immediately follows it operates on. It seems to me that if there _is_ a current lease, then the lease holder is immediately considered for lease rebalancing, and that's all that this comment really needs to say.
> 
> Also, is it possible to get here without holding the lease? 
</details>
Yes, we can get here without holding the lease. We only grab the lease for calling `process()`. But we only get here if we're the leaseholder or there is no leaseholder.

I've updated the comment and also tweaked the code so we only call `ShouldTransferLease` if we're currently holding the lease.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r12.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replicate_queue.go, line 118 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KWZ-8VnvtRNN6UIIfIr:-KWZ-8Vog88ai0tjZDVH:bu9kllo) ([raw file](https://github.com/cockroachdb/cockroach/blob/eef3cdec09e292caed4dd8c8b78be4922dc240af/pkg/storage/replicate_queue.go#L118)):*
> ```Go
> 	}
> 
> 	// If we hold the lease check to see if we should transfer for it.
> ```

s/for// ? also add a comma for the subordinate clause.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 4 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replicate_queue.go, line 118 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464#-KWZ-8VnvtRNN6UIIfIr:-KWZ0UrkaMzZPLhAMUc8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eef3cdec09e292caed4dd8c8b78be4922dc240af/pkg/storage/replicate_queue.go#L118)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/for// ? also add a comma for the subordinate clause.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r13.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10464)*
<!-- Sent from Reviewable.io -->
